### PR TITLE
Include outer join rels in Join Order algorithms

### PR DESCRIPTION
**Pull request by sambitesh**
This commit achieves two purposes:

1) `Exhaustive` join order now includes `Greedy` too. Earlier `Greedy` would be
    disabled for `Exhaustive` search.

2) Outer join rels are now included with NaryJoin rels in `Greedy` and
  `MinCard` join order algorithms. `JoinOrderDP` is not changed.  As a result, a
  transormation like below is possible now (through `Greedy` and `Mincard`).

    Nary              
     |_t1                 
     |_LOJ   
     .  . |_t2              
     .  . |_t3   
           
  to below

LOJ
  |_Nary
   | . .  |_t1
   | . .  |_t2
   |_t3  

We track rels of outer joins with the help of their `SComponents`. We
introduced `outerchild_index` and `innerchild_index` to `SComponent`. While
joining two rels we make sure that their `outerchild_index` and
`innerchild_index` are compatible.

Signed-off-by: Dhanashree Kashid <dkashid@pivotal.io>

### CODE REVIEW

File: libgpopt/include/gpopt/xforms/CXformExpandNAryJoinGreedy.h [hunk 1]
```diff
@@ -81,7 +81,7 @@ namespace gpopt
 					CXformResult *pxfres,
 					CExpression *pexpr
 					) const;
-
+BOOL IsApplyOnce() { return true; }
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/xforms/CXformExpandNAryJoinGreedy.h]: why all of a sudden? Should we run performance pipeline for this?

Comment (vraghavan78) [reply libgpopt/include/gpopt/xforms/CXformExpandNAryJoinGreedy.h]: > why all of a sudden? Should we run performance pipeline for this?

I see now why. But we should run performance pipeline with and without partitions

File: libgpopt/src/operators/CNormalizer.cpp [hunk 2]
```diff
@@ -145,6 +145,51 @@ CNormalizer::FPushable
...
+//---------------------------------------------------------------------------
+//	@function:
+//		CNormalizer::FPushable
+//
+//	@doc:
+//		Check if a predicate is part of current Left Outer Join's
+//		join condition.
+//
+//
+//---------------------------------------------------------------------------
```
Comment (vraghavan78) [inline libgpopt/src/operators/CNormalizer.cpp]: lets remove the paddings. Clean house as we go along.

File: libgpopt/src/operators/CNormalizer.cpp [hunk 3]
```diff
@@ -145,6 +145,51 @@ CNormalizer::FPushable
...
+BOOL
+CNormalizer::IsPredAnLOJPred
+	(
+	IMemoryPool *mp,
+	CExpression *pexprLogical,
+	CExpression *pexprPred
+	)
+{
+	GPOS_ASSERT(NULL != pexprLogical);
+	GPOS_ASSERT(NULL != pexprPred);
```
Comment (vraghavan78) [inline libgpopt/src/operators/CNormalizer.cpp]: shouldn't we check if it is a LOJ?

File: libgpopt/src/operators/CNormalizer.cpp [hunk 4]
```diff
@@ -456,7 +501,7 @@ CNormalizer::PushThruSelect
 		CExpressionArray *pdrgpexprConjuncts = CPredicateUtils::PdrgpexprConjuncts(mp, pexprPred);
 		CExpressionArray *pdrgpexprRemaining = NULL;
 		CExpression *pexpr = NULL;
-		PushThru(mp, pexprLogicalChild, pdrgpexprConjuncts, &pexpr, &pdrgpexprRemaining);
+		PushThru(mp, pexprLogicalChild, pdrgpexprConjuncts, &pexpr, &pdrgpexprRemaining, false);
```
Comment (vraghavan78) [inline libgpopt/src/operators/CNormalizer.cpp]: why false? Can add a comment in /* */

File: libgpopt/src/operators/CNormalizer.cpp [hunk 5]
```diff
@@ -557,7 +602,7 @@ CNormalizer::PushThruUnaryWithoutScalarChild
 	// push conjuncts through the logical child
 	CExpression *pexprNewLogicalChild = NULL;
 	CExpressionArray *pdrgpexprUnpushable = NULL;
-	PushThru(mp, pexprLogicalChild, pdrgpexprConjuncts, &pexprNewLogicalChild, &pdrgpexprUnpushable);
+	PushThru(mp, pexprLogicalChild, pdrgpexprConjuncts, &pexprNewLogicalChild, &pdrgpexprUnpushable, false);
```
Comment (vraghavan78) [inline libgpopt/src/operators/CNormalizer.cpp]: same here

File: libgpopt/src/operators/CNormalizer.cpp [hunk 6]
```diff
@@ -990,13 +1037,22 @@ CNormalizer::PushThru
...
 	const ULONG size = pdrgpexprConjuncts->Size();
 	for (ULONG ul = 0; ul < size; ul++)
 	{
 		CExpression *pexprConj = (*pdrgpexprConjuncts)[ul];
 		pexprConj->AddRef();
-		if (FPushable(pexprLogical, pexprConj))
+
+		if (isLOJWithInnerJoinParent)
+		{
+			pushable = IsPredAnLOJPred(mp, pexprLogical, pexprConj);
```
Comment (vraghavan78) [inline libgpopt/src/operators/CNormalizer.cpp]: isLOJWithInnerJoinParent should be inside the function.

File: libgpopt/src/operators/CNormalizer.cpp [hunk 7]
```diff
@@ -145,6 +145,51 @@ CNormalizer::FPushable
...
+//		CNormalizer::FPushable
+//
+//	@doc:
+//		Check if a predicate is part of current Left Outer Join's
+//		join condition.
+//
+//
+//---------------------------------------------------------------------------
+BOOL
+CNormalizer::IsPredAnLOJPred
```
Comment (hardikar) [inline libgpopt/src/operators/CNormalizer.cpp]: `IsPredAnLOJPred` doesn't really describe the what this function is doing.

File: libgpopt/src/operators/CNormalizer.cpp [hunk 8]
```diff
@@ -1010,7 +1066,22 @@ CNormalizer::PushThru
 	CExpression *pexprPred = CPredicateUtils::PexprConjunction(mp, pdrgpexprPushable);
 	if (FPushThruOuterChild(pexprLogical))
 	{
-		PushThruOuterChild(mp, pexprLogical, pexprPred, ppexprResult);
+		// if the parent of this LOJ is an inner join, then we don't care about
+		// pushing predicates down LOJ's outer children
+		if (is_child_of_inner_join || (0 == pexprLogical->Arity()))
```
Comment (hardikar) [inline libgpopt/src/operators/CNormalizer.cpp]: Can't this check be somehow incorporated in FPushThruOuterChild? Similarly earlier in this function using FPushable?

File: libgpopt/src/operators/CNormalizer.cpp [hunk 9]
```diff
@@ -857,7 +903,7 @@ CNormalizer::PushThruJoin
 		}
 
 		CExpressionArray *pdrgpexprRemaining = NULL;
-		PushThru(mp, pexprChild, pdrgpexprConjuncts, &pexprNewChild, &pdrgpexprRemaining);
+		PushThru(mp, pexprChild, pdrgpexprConjuncts, &pexprNewChild, &pdrgpexprRemaining, is_child_of_inner_join);
```
Comment (hardikar) [inline libgpopt/src/operators/CNormalizer.cpp]: I don't really like the precedent of passing bool values like `is_child_of_inner_join`. This method will start growing bigger as we consider every possible parent type as an additional parameter.

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 10]
```diff
@@ -214,29 +243,79 @@ CJoinOrder::CJoinOrder
...
 	m_rgpcomp(NULL),
-	m_ulComps(0)
+	m_ulComps(0),
+	m_include_outer_join_rels (include_outer_join_rels)
 {
 	typedef SComponent* Pcomp;
 	typedef SEdge* Pedge;
 	
-	m_ulComps = pdrgpexpr->Size();
+	ULONG nary_children = pdrgpexpr->Size();
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CJoinOrder.cpp]: const

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 11]
```diff
@@ -214,29 +243,79 @@ CJoinOrder::CJoinOrder
...
+	INT outerJoins = 0;
+
+	if (m_include_outer_join_rels)
+	{
+		for (ULONG ul = 0; ul < nary_children; ul++)
+		{
+			CExpression *pexprComp = (*pdrgpexpr)[ul];
+			if (COperator::EopLogicalLeftOuterJoin == pexprComp->Pop()->Eopid())
+			{
+				// TODO: handle nested case here
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CJoinOrder.cpp]: can we instead state above, we handle only one level of LOJ

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 12]
```diff
@@ -214,29 +243,79 @@ CJoinOrder::CJoinOrder
...
+				outerJoins++;
+			}
+		}
+	}
+
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CJoinOrder.cpp]: can we have some comments that explain what is going on. It is very hard to understand otherwise

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 13]
```diff
@@ -214,29 +243,79 @@ CJoinOrder::CJoinOrder
 	(
 	IMemoryPool *mp,
 	CExpressionArray *pdrgpexpr,
-	CExpressionArray *pdrgpexprConj
+	CExpressionArray *pdrgpexprConj,
+	BOOL include_outer_join_rels
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CJoinOrder.cpp]: why do we need this boolean?

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 14]
```diff
@@ -214,29 +243,79 @@ CJoinOrder::CJoinOrder
...
+		for (ULONG ul = 0; ul < nary_children; ul++)
+		{
+			CExpression *pexprComp = (*pdrgpexpr)[ul];
+			if (COperator::EopLogicalLeftOuterJoin == pexprComp->Pop()->Eopid())
+			{
+				// TODO: handle nested case here
+				outerJoins++;
+			}
+		}
+	}
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CJoinOrder.cpp]: the above just iterates and tell you how many LOJ are there.

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 15]
```diff
@@ -214,29 +243,79 @@ CJoinOrder::CJoinOrder
...
+	BOOL include_outer_join_rels
 	)
 	:
 	m_mp(mp),
 	m_rgpedge(NULL),
 	m_ulEdges(0),
 	m_rgpcomp(NULL),
-	m_ulComps(0)
+	m_ulComps(0),
+	m_include_outer_join_rels (include_outer_join_rels)
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CJoinOrder.cpp]: you not handle FULLOUTER joins just LOJ, so the name is confusing.

File: libgpopt/src/xforms/CJoinOrder.cpp (hunk 14 above)
Comment (vraghavan78) [reply libgpopt/src/xforms/CJoinOrder.cpp]: why are your iterating over this twice?

Comment (vraghavan78) [reply libgpopt/src/xforms/CJoinOrder.cpp]: nevermind, i see why u need to

File: libgpopt/src/xforms/CJoinOrder.cpp (hunk 13 above)
Comment (sambitesh) [reply libgpopt/src/xforms/CJoinOrder.cpp]: DP won't be including the outer join rels but MinCard and Greedy will be. So we extract outer join rels based on this boolean value

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 16]
```diff
@@ -219,29 +252,101 @@ CJoinOrder::CJoinOrder
...
+	// |  |--CLogicalGet "t5"
+	// |  |--CLogicalGet "t4"
+	// |  +--CScalarCmp (=)
+	// |
+	// +--CScalarCmp (=)
+	//
+	// In above case the pdrgpexpr comes with two elements in it:
+	//  - CLogicalGet "t1"
+	//  - CLogicalLeftOuterJoin
+	// We need to create compontnents out of "t1", "t4", "t5" and store them
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CJoinOrder.cpp]: components

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 17]
```diff
@@ -219,29 +252,101 @@ CJoinOrder::CJoinOrder
...
+
+	if (m_include_left_outer_join_rels)
+	{
+		for (ULONG ul = 0; ul < num_of_nary_children; ul++)
+		{
+			CExpression *pexprComp = (*pdrgpexpr)[ul];
+			if (COperator::EopLogicalLeftOuterJoin == pexprComp->Pop()->Eopid())
+			{
+				// TODO: we handle only one level of LOJ
+				// need to handle nested case too
```
Comment (vraghavan78) [inline libgpopt/src/xforms/CJoinOrder.cpp]: why is this a TODO. Instead state we currently only one level of LOJ

File: libgpopt/include/gpopt/xforms/CXformExpandNAryJoinGreedy.h (hunk 1 above)
Comment (bhuvnesh2703) [reply libgpopt/include/gpopt/xforms/CXformExpandNAryJoinGreedy.h]: ran 256 gb non partitioned perf pipeline, no regressions.

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 18]
```diff
@@ -214,29 +243,102 @@ CJoinOrder::CJoinOrder
...
+			CExpression *outer_child = (*pexprComp)[0];
+			CExpression *inner_child = (*pexprComp)[1];
+
+			outer_child->AddRef();
+			SComponent *sc_outer = GPOS_NEW(mp) SComponent(mp, outer_child);
+			// the outerchild and inner_child of an LOJ get an outerchild_index and innerchild_index for the respective outer
+			// and inner child.
+			// for same LOJ, the outerchild_index = innerchild_index
+			// different LOJs get different indexes
+			sc_outer->SetOuterChildIndex(++outerchild_index);
```
[the remark this answers is not in the document]
Comment (bhuvnesh2703) [reply libgpopt/src/xforms/CJoinOrder.cpp]: didn't consider this change.

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 19]
```diff
@@ -448,5 +582,167 @@ CJoinOrder::OsPrint
...
+
+#ifdef GPOS_DEBUG
+		CAutoRef<CBitSet> pbsScalarConjuctsMatched(GPOS_NEW(m_mp) CBitSet(m_mp));
+#endif
+		ULONG ulMatchCount = 0; // Count of edge predicate conjucts matched
+		// For each conjuct of the edge predicate
+		for (ULONG ulEdgePred = 0; ulEdgePred < ulSizeEdge; ++ulEdgePred)
+		{
+			// For each conjuct of the scalar predicate
+			for (ULONG ulScalarPred = 0; ulScalarPred < ulSizeScalar; ulScalarPred++)
```
Comment (hardikar) [inline libgpopt/src/xforms/CJoinOrder.cpp]: Can we not use SComponent:: m_edge_set here?

Comment (sambitesh) [reply libgpopt/src/xforms/CJoinOrder.cpp]: @hardikar done - take another look

Comment (hardikar) [reply libgpopt/src/xforms/CJoinOrder.cpp]: Hold on, why do we need to still iterate over all the predicates?

Comment (sambitesh) [reply libgpopt/src/xforms/CJoinOrder.cpp]: I don't get you question? We still get hyper-edges like before. That's why I have to loop through all predicates.

File: libgpopt/include/gpopt/xforms/CXformExpandNAryJoin.h [hunk 20]
```diff
@@ -71,7 +71,7 @@ namespace gpopt
 					CExpression *pexpr
 					) const;
 
-	}; // class CXformExpandNAryJoin
+				}; // class CXformExpandNAryJoin
```
Comment (hardikar) [inline libgpopt/include/gpopt/xforms/CXformExpandNAryJoin.h]: Wrong indentation?

File: libgpopt/include/gpopt/xforms/CJoinOrder.h [hunk 21]
```diff
@@ -83,6 +83,13 @@ namespace gpopt
 				// a flag to component edge as used
 				BOOL m_fUsed;
 
+				// populated only for left outer join child components
+				// it is an incremental counter for the LOJ's available in tree,
+				// and for children of same LOJ, m_outerchild_index == m_innerchild_index
+				INT m_outerchild_index;
+
+				INT m_innerchild_index;
```
Comment (hardikar) [inline libgpopt/include/gpopt/xforms/CJoinOrder.h]:
```suggestion
				// parent LOJ whose inner child is this component
				INT m_innerchild_index;
```

File: libgpopt/include/gpopt/xforms/CJoinOrder.h [hunk 22]
```diff
@@ -83,6 +83,13 @@ namespace gpopt
 				// a flag to component edge as used
 				BOOL m_fUsed;
 
+				// populated only for left outer join child components
+				// it is an incremental counter for the LOJ's available in tree,
+				// and for children of same LOJ, m_outerchild_index == m_innerchild_index
+				INT m_outerchild_index;
```
Comment (hardikar) [inline libgpopt/include/gpopt/xforms/CJoinOrder.h]:
```suggestion
				// parent LOJ whose outer child is this component
				INT m_outerchild_index;
```

File: libgpopt/include/gpopt/xforms/CJoinOrder.h [hunk 23]
```diff
@@ -83,6 +83,13 @@ namespace gpopt
 				// a flag to component edge as used
 				BOOL m_fUsed;
 
+				// populated only for left outer join child components
+				// it is an incremental counter for the LOJ's available in tree,
```
Comment (hardikar) [inline libgpopt/include/gpopt/xforms/CJoinOrder.h]:
```suggestion
				// it is an index into the list of LOJ's available in join tree.
```

File: libgpopt/include/gpopt/xforms/CJoinOrder.h [hunk 24]
```diff
@@ -83,6 +83,13 @@ namespace gpopt
...
+				// and for children of same LOJ, m_outerchild_index == m_innerchild_index
```
Comment (hardikar) [inline libgpopt/include/gpopt/xforms/CJoinOrder.h]:
```suggestion
				// and for children components (s1, s2) of the same LOJ,  s1.m_outerchild_index == s2.m_innerchild_index
```

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 25]
```diff
@@ -367,21 +471,51 @@ CJoinOrder::PcompCombine
...
 		// first call to this function, we create a Select node
+		component_outerchild_index = pcompInner->GetOuterChildIndex();
+		component_innerchild_index = pcompInner->GetInnerChildIndex();
 		pexpr = CUtils::PexprCollapseSelect(m_mp, pexprInner, pexprScalar);
 		pexprScalar->Release();
 	}
 	else
 	{
-		// not first call, we create an Inner Join
 		pexprInner->AddRef();
```
Comment (hardikar) [inline libgpopt/src/xforms/CJoinOrder.cpp]:
```suggestion
		GPOS_ASSERT(NULL != pexprInner);
		pexprInner->AddRef();
```

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 26]
```diff
@@ -367,21 +471,51 @@ CJoinOrder::PcompCombine
...
+		{
+			pexpr = CUtils::PexprLogicalJoin<CLogicalLeftOuterJoin>(m_mp, pexprOuter, pexprInner, pexprScalar);
+		}
+		else if (IsChildOfSameLOJ(pcompInner, pcompOuter))
+		{
+			pexpr = CUtils::PexprLogicalJoin<CLogicalLeftOuterJoin>(m_mp, pexprInner, pexprOuter, pexprScalar);
+		}
+		else
+		{
+			if (pcompOuter->GetOuterChildIndex() > 0)
```
Comment (hardikar) [inline libgpopt/src/xforms/CJoinOrder.cpp]: Is there a reason why we don't need to update component_innerchild_index?

For example if we have (t1 LOJ t2) INNER t3, if PcompCombine(t2, t3) is called, the resulting component would have both inner & outer child indexes set to 0, right? Is that expected?

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 27]
```diff
@@ -448,5 +582,123 @@ CJoinOrder::OsPrint
...
 
+	// check if both the components are a part of an LOJ (not necessary the same LOJ), else
+	// we do not want to continue with this combination
+	if ((component_1->GetOuterChildIndex() == 0 && component_2->GetInnerChildIndex() > 0) ||
+		(component_2->GetOuterChildIndex() == 0 && component_1->GetInnerChildIndex() > 0))
+	{
+		return false;
+	}
+
+	return true;
```
Comment (hardikar) [inline libgpopt/src/xforms/CJoinOrder.cpp]: if component1 and component2 are have both inner & outer as 0, should IsValidLOJCombination() return 0?

Comment (hardikar) [reply libgpopt/src/xforms/CJoinOrder.cpp]: Maybe it's just that the function needs to be renamed.

File: libgpopt/src/xforms/CJoinOrder.cpp (hunk 26 above)
Comment (dhanashreek89) [reply libgpopt/src/xforms/CJoinOrder.cpp]: The innerchild index of component is never updated, only the outer index is updated when it is combined with other component. Inner child index is used to find the right outer component for it.

In the example, not sure if (t2, t3) is a valid combination to start with. The caller of this function should already handle that via `IsValidLOJCombination()`.

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 28]
```diff
@@ -448,5 +582,123 @@ CJoinOrder::OsPrint
...
+
+	while (edges_iter.Advance())
+	{
+		SEdge *pedge = m_rgpedge[edges_iter.Bit()];
+		if (pedge->m_fUsed)
+		{
+			continue;
+		}
+
+		if (pcomponent->m_pbs->ContainsAll(pedge->m_pbs))
```
[the remark this answers is not in the document]
Comment (sambitesh) [reply libgpopt/src/xforms/CJoinOrder.cpp]: None of the join order algorithms will give you a hashjoin for t1 = t2 + t3 today. We have filed a story to fix it.

File: libgpopt/src/xforms/CXformExpandNAryJoinGreedy.cpp [hunk 29]
```diff
@@ -47,7 +45,7 @@ CXformExpandNAryJoinGreedy::CXformExpandNAryJoinGreedy
 					(
 					pmp,
 					GPOS_NEW(pmp) CLogicalNAryJoin(pmp),
-					GPOS_NEW(pmp) CExpression(pmp, GPOS_NEW(pmp) CPatternMultiLeaf(pmp)),
+					GPOS_NEW(pmp) CExpression(pmp, GPOS_NEW(pmp) CPatternMultiTree(pmp)),
```
Comment (zellerh) [inline libgpopt/src/xforms/CXformExpandNAryJoinGreedy.cpp]: Sorry, I'm still new in this project, but from my experience with Cascades, here is a question: If we materialize the tree all the way down to the leaves, could this cause an explosion? Here is an example: Let's say we have an NAry join with an LOJ underneath, and that LOJ has a very complex child, let's say a groupby on top of a 10-way join. There could be a billion different ways to materialize a binding for this complex child. Does this make sense or is there something that limits the number of different trees we materialize to do the xform?

Comment (sambitesh) [reply libgpopt/src/xforms/CXformExpandNAryJoinGreedy.cpp]: You are right. That's why we have a flag called `FApplyIOnce` which limits the different number of trees we can generate.

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 30]
```diff
@@ -367,21 +471,52 @@ CJoinOrder::PcompCombine
...
 	{
-		// not first call, we create an Inner Join
+		GPOS_ASSERT(NULL != pexprInner);
 		pexprInner->AddRef();
 		pexprOuter->AddRef();
-		pexpr = CUtils::PexprLogicalJoin<CLogicalInnerJoin>(m_mp, pexprOuter, pexprInner, pexprScalar);
+
+		if (IsChildOfSameLOJ(pcompOuter, pcompInner))
+		{
+			pexpr = CUtils::PexprLogicalJoin<CLogicalLeftOuterJoin>(m_mp, pexprOuter, pexprInner, pexprScalar);
```
Comment (zellerh) [inline libgpopt/src/xforms/CJoinOrder.cpp]: Should we set component_innerchild_index here, since we are dealing with a component that has the inner of an LOJ?

Comment (sambitesh) [reply libgpopt/src/xforms/CJoinOrder.cpp]: Not in this case. This component has an LOJ expression. If we set the component_innerchild_index here, it will fail the `isValidLOJ` check later when combining with other components.

File: data/dxl/minidump/NaryJoinWithLOJasChild.mdp [hunk 31]
```diff
@@ -0,0 +1,2916 @@
...
+	 CREATE TABLE test4 ( a INT, b INT);
+	 CREATE TABLE test5 ( a INT, b INT);
+	 INSERT INTO test1 SELECT i,i from generate_series(1,100)i;
+	 INSERT INTO test4 SELECT i,i from generate_series(1,100000)i;
+	 INSERT INTO test5 SELECT i,i from generate_series(1,1000000)i;
+	 Objective: For a query with outerjoins and inner joins, let the rels in outerjoins
+	 participate in join order along with rels of inner joins. 
+
+	 EXPLAIN SELECT * 
+	 FROM   t1
```
Comment (bhuvnesh2703) [inline data/dxl/minidump/NaryJoinWithLOJasChild.mdp]: pleaes update the table DDL? t1 or test1?

File: libgpopt/src/xforms/CJoinOrderGreedy.cpp [hunk 32]
```diff
@@ -276,6 +239,11 @@ CJoinOrderGreedy::PickBestJoin
 		SComponent *pcompCurrent = m_rgpcomp[iter.Bit()];
 		SComponent *pcompTemp = PcompCombine(m_pcompResult, pcompCurrent);
 
+		if (!IsValidLOJCombination(m_pcompResult, pcompCurrent))
```
Comment (bhuvnesh2703) [inline libgpopt/src/xforms/CJoinOrderGreedy.cpp]: continue here will cause leak, it should go above `PcompCombine`.. pcompTemp should be freed if we continue.. or better call PcompCombine only after checking the combination..
We should put an assert in PcompCombine to validate if the Join combination is valid.

File: libgpopt/src/xforms/CJoinOrder.cpp [hunk 33]
```diff
@@ -448,5 +583,123 @@ CJoinOrder::OsPrint
...
+
+	while (edges_iter.Advance())
+	{
+		SEdge *pedge = m_rgpedge[edges_iter.Bit()];
+		if (pedge->m_fUsed)
+		{
+			continue;
+		}
+
+		if (pcomponent->m_pbs->ContainsAll(pedge->m_pbs))
```
Comment (zellerh) [inline libgpopt/src/xforms/CJoinOrder.cpp]: This method uses the containment test to create edges between components. With LOJs, there are additional things we need to consider. For example, if we have this join:

```
select * from T1 cross join T2 left join T3 on t1.b=t2.b
```

This algorithm would assign the predicate t1.b=t2.b to the component that contains { T1, T2 }, but it really belongs to a component that contains { T1, T2, T3 } and it needs to be evaluated as the join predicate of the LOJ.

I don't see wrong results, probably because we use SComponent::m_pexpr to create the result expression, and that data member has the predicates in the right place.

Comment (bhuvnesh2703) [reply libgpopt/src/xforms/CJoinOrder.cpp]: For the query `select * from T1 cross join T2 left join T3 on t1.b=t2.b`, Greedy Xform will not be triggered for the tree starting with `CLogicalLeftOuterJoin`..It is only triggered for `CLogicalNAryJoin`. Refer to the PxfsCoandidates for `CLogicalLeftOuterJoin` and `CLogicalNAryJoin` for the list of Xform triggered.

Here is the tree for the query:
```
+--CLogicalLeftOuterJoin
   |--CLogicalNAryJoin
   |  |--CLogicalGet "t1" ("t1"), Columns: ["a" (0), "b" (1), "ctid" (2), "xmin" (3), "cmin" (4), "xmax" (5), "cmax" (6), "tableoid" (7), "gp_segment_id" (8)] Key sets: {[2,8]}
   |  |--CLogicalGet "t2" ("t2"), Columns: ["c" (9), "d" (10), "ctid" (11), "xmin" (12), "cmin" (13), "xmax" (14), "cmax" (15), "tableoid" (16), "gp_segment_id" (17)] Key sets: {[2,8]}
   |  +--CScalarConst (1)
   |--CLogicalGet "t3" ("t3"), Columns: ["e" (18), "f" (19), "ctid" (20), "xmin" (21), "cmin" (22), "xmax" (23), "cmax" (24), "tableoid" (25), "gp_segment_id" (26)] Key sets: {[2,8]}
   +--CScalarCmp (=)
      |--CScalarIdent "b" (1)
      +--CScalarIdent "d" (10)
```
Ok, let's use the modified version of the query just for triggering Greedy:
`explain select * from (T1 cross join T2 left join T3 on t1.b=t2.d) t4 inner join t2 on t4.b = t2.c`
Here is the tree for it.
```
+--CLogicalNAryJoin
   |--CLogicalLeftOuterJoin
   |  |--CLogicalNAryJoin
   |  |  |--CLogicalGet "t1" ("t1"), Columns: ["a" (0), "b" (1), "ctid" (2), "xmin" (3), "cmin" (4), "xmax" (5), "cmax" (6), "tableoid" (7), "gp_segment_id" (8)] Key sets: {[2,8]}
   |  |  |--CLogicalGet "t2" ("t2"), Columns: ["c" (9), "d" (10), "ctid" (11), "xmin" (12), "cmin" (13), "xmax" (14), "cmax" (15), "tableoid" (16), "gp_segment_id" (17)] Key sets: {[2,8]}
   |  |  +--CScalarConst (1)
   |  |--CLogicalGet "t3" ("t3"), Columns: ["e" (18), "f" (19), "ctid" (20), "xmin" (21), "cmin" (22), "xmax" (23), "cmax" (24), "tableoid" (25), "gp_segment_id" (26)] Key sets: {[2,8]}
   |  +--CScalarCmp (=)
   |     |--CScalarIdent "b" (1)
   |     +--CScalarIdent "d" (10)
   |--CLogicalGet "t4" ("t2"), Columns: ["c" (27), "d" (28), "ctid" (29), "xmin" (30), "cmin" (31), "xmax" (32), "cmax" (33), "tableoid" (34), "gp_segment_id" (35)] Key sets: {[2,8]}
   +--CScalarCmp (=)
      |--CScalarIdent "b" (1)
      +--CScalarIdent "c" (27)
```
When we will create components for `CLogicalLeftOuterJoin`
comp1 =  CLogicalNAryJoin = we will assign this as the left component not t1 or t2
comp2 = CLogicalGet "t4"
scalar condition = which joins `CLogicalLeftOuterJoin`i.e comp1 and `CLogicalGet "t4"` and comp2
```
   +--CScalarCmp (=)
      |--CScalarIdent "b" (1)
      +--CScalarIdent "c" (27)
```
When containsall will check it will check whether this edge covers `CLogicalLeftOuterJoin`i.e comp1 and `CLogicalGet "t4"` {instead of checking if t1, t2 NaryJoin are covered by scalar cmp)
so, in other words the condition will be attached to the LOJ component and so we will not assign the predicates incorrectly.

**Comment by bhuvnesh2703** [comment]
Took the feedback and created a new PR. Also, did little refactoring.. For your review: https://github.com/greenplum-db/gporca/pull/424
